virtool/users/transforms.py

```
@@ -0,0 +1,97 @@
+    ) -> List[Document]:
+        attached = []
+
+        for document in documents:
+            attached.append(
+                {
+                    **document,
+                    "permissions": merge_group_permissions(
+                        [
+                            prepared[group_id]
+                            for group_id in self.get_group_ids(document)
+                        ]
+                    ),
+                }
+            )
+
+        return attached
+
+    async def prepare_one(self, document) -> Optional[Document]:
+        group_ids = self.get_group_ids(document)
+        return await self.get_groups_from_ids(list(group_ids))
+
+    async def prepare_many(
+        self, documents: List[Document]
+    ) -> Dict[Union[int, str], Any]:
+        group_ids = set()
+
+        for user in documents:
+            group_ids.update(self.get_group_ids(user))
+
+        return await self.get_groups_from_ids(list(group_ids))
+
+    async def get_groups_from_ids(self, group_ids: list[str]):
+        if len(group_ids) == 0:
+            return {}
+
+        async with both_transactions(self._mongo, self._pg) as (
+            mongo_session,
+            pg_session,
+        ):
+
+            mongo_result, pg_result = await asyncio.gather(
+                self._mongo.groups.find(
+                    {"_id": {"$in": group_ids}}, session=mongo_session
+                ).to_list(None),
+                pg_session.execute(
+                    select(SQLGroup).filter(SQLGroup.legacy_id.in_(group_ids))
+                ),
+            )
+
+            mongo_groups = {group["_id"]: group for group in mongo_result}
+            pg_groups = {
+                group.legacy_id: group.to_dict() for group in pg_result.scalars()
+            }
+
+            return mongo_groups | pg_groups
+
+    @staticmethod
+    def get_group_ids(user) -> set[str]:
+        groups = get_safely(user, "groups")
```

Not sure this is something that we should let fail silently. When are you going to be applying the transform but some documents will by missing the `groups` field?